flow/raster_cache_key.h

```diff
@@ -15,60 +15,77 @@
 
 namespace flutter {
 
+enum class RasterCacheKeyType {
+  kLayer,
+  kPicture,
+  kDisplayList,
+  kLayerChildren
+};
+
+enum class RasterCacheKeyKind { kLayerMetrics, kPictureMetrics };
+
 class RasterCacheKeyID {
  public:
-  RasterCacheKeyID(const std::vector<uint64_t> ids) : ids_(ids) {}
+  static constexpr uint64_t kDefaultUniqueID = 0;
+
+  RasterCacheKeyID(uint64_t unique_id, RasterCacheKeyType type)
+      : unique_id_(unique_id), type_(type) {}
 
-  const std::vector<uint64_t>& ids() { return ids_; }
+  RasterCacheKeyID(std::vector<RasterCacheKeyID> child_ids,
+                   RasterCacheKeyType type)
+      : unique_id_(kDefaultUniqueID),
+        type_(type),
+        child_ids_(std::move(child_ids)) {}
+
+  uint64_t unique_id() const { return unique_id_; }
+
+  RasterCacheKeyType type() const { return type_; }
+
+  const std::vector<RasterCacheKeyID>& child_ids() const { return child_ids_; }
 
   std::size_t GetHash() const {
     std::size_t seed = fml::HashCombine();
-    for (auto id : ids_) {
-      fml::HashCombineSeed(seed, id);
+    fml::HashCombineSeed(seed, unique_id_);
+    fml::HashCombineSeed(seed, type_);
+    for (auto& child_id : child_ids_) {
+      fml::HashCombineSeed(seed, child_id.GetHash());
     }
     return seed;
   }
 
   bool operator==(const RasterCacheKeyID& other) const {
-    return ids_ == other.ids_;
+    return unique_id_ == other.unique_id_ && type_ == other.type_ &&
+           child_ids_ == other.child_ids_;
```

Comment:
If we cache the hash, then it might save time to compare before doing the child_ids_ vector compare.

Reply:
done